File: backend/services/auth_service.py
import jwt
import requests
from datetime import datetime, timedelta
from repositories.user_repository import UserRepository
from models.user import User
import os

class AuthService:
    JWT_SECRET = os.getenv('JWT_SECRET', 'your-secret-key-change-this')
    JWT_EXPIRATION_HOURS = 24
    
    @classmethod
    def generate_jwt_token(cls, user_id):
        """Generate JWT token for user"""
        payload = {
            'user_id': str(user_id),
            'exp': datetime.utcnow() + timedelta(hours=cls.JWT_EXPIRATION_HOURS),
            'iat': datetime.utcnow()
        }
        return jwt.encode(payload, cls.JWT_SECRET, algorithm='HS256')
    
    @classmethod
    def verify_jwt_token(cls, token):
        """Verify JWT token and return user ID"""
        try:
            data = jwt.decode(token, cls.JWT_SECRET, algorithms=['HS256'])
            return data['user_id']
        except jwt.ExpiredSignatureError:
            raise ValueError('Token has expired')
        except jwt.InvalidTokenError:
            raise ValueError('Token is invalid')
    
    @classmethod
    def register_user(cls, name, email, password, business_info=None):
        """Register a new user"""
        # Validate input
        if not name or not name.strip():
            raise ValueError('Name is required')
        if not email or not email.strip():
            raise ValueError('Email is required')
        if not password:
            raise ValueError('Password is required')
        
        email = email.strip().lower()
        name = name.strip()
        
        if not User.validate_email(email):
            raise ValueError('Invalid email format')
        
        is_valid, message = User.validate_password(password)
        if not is_valid:
            raise ValueError(message)
        
        # Check if user exists
        existing_user = UserRepository.find_by_email(email)
        if existing_user:
            raise ValueError('User with this email already exists')
        
        # Create user
        user_data = {
            'email': email,
            'name': name,
            'provider': 'email',
            'is_active': True,
            'business_info': business_info or {
                'business_name': name,
                'business_type': 'Du lịch',
                'industry': 'Tourism',
                'phone': '',
                'address': ''
            }
        }
        
        user = User(user_data)
        user.set_password(password)
        
        # Save to database
        created_user = UserRepository.create_user(user.to_db_dict())
        
        # Generate token
        token = cls.generate_jwt_token(created_user._id)
        
        return {
            'user': created_user.to_dict(),
            'token': token
        }
    
    @classmethod
    def login_user(cls, email, password):
        """Login user with email and password"""
        if not email or not email.strip():
            raise ValueError('Email is required')
        if not password:
            raise ValueError('Password is required')
        
        email = email.strip().lower()
        
        # Find user
        user = UserRepository.find_by_email(email)
        if not user:
            raise ValueError('Invalid email or password')
        
        # Check if user registered with social login
        if user.provider != 'email':
            raise ValueError(f'This account was created using {user.provider.title()} login. Please use {user.provider.title()} to sign in.')
        
        # Check password
        if not user.check_password(password):
            raise ValueError('Invalid email or password')
        
        # Check if user is active
        if not user.is_active:
            raise ValueError('Account is deactivated')
        
        # Update last login
        UserRepository.update_last_login(user._id)
        
        # Generate token
        token = cls.generate_jwt_token(user._id)
        
        return {
            'user': user.to_dict(),
            'token': token
        }
    
    @classmethod
    def google_login(cls, google_token):
        """Login with Google OAuth"""
        try:
            # Verify Google ID token
            google_client_id = os.getenv('GOOGLE_CLIENT_ID')
            if not google_client_id:
                raise ValueError('Google authentication not configured')
            
            # Verify ID token with Google
            verify_response = requests.get(
                f'https://oauth2.googleapis.com/tokeninfo?id_token={google_token}'
            )
            
            if verify_response.status_code != 200:
                raise ValueError('Invalid Google token')
            
            user_info = verify_response.json()
            
            # Verify audience (client ID)
            if user_info.get('aud') != google_client_id:
                raise ValueError('Invalid token audience')
            
            email = user_info.get('email', '').lower()
            name = user_info.get('name', '')
            google_id = user_info.get('sub')
            profile_picture = user_info.get('picture')
            
            if not email or not google_id:
                raise ValueError('Invalid Google user data')
            
            # Check if user exists
            user = UserRepository.find_by_email(email)
            
            if user:
                # Update existing user
                update_data = {
                    'google_id': google_id,
                    'profile_picture': profile_picture,
                    'last_login': datetime.utcnow()
                }
                UserRepository.update_user(user._id, update_data)
                
                # Refresh user data
                user = UserRepository.find_by_id(user._id)
            else:
                # Create new user
                user_data = {
                    'email': email,
                    'name': name,
                    'google_id': google_id,
                    'provider': 'google',
                    'profile_picture': profile_picture,
                    'is_active': True,
                    'last_login': datetime.utcnow(),
                    'business_info': {
                        'business_name': name,
                        'business_type': 'Du lịch',
                        'industry': 'Tourism',
                        'phone': '',
                        'address': ''
                    }
                }
                
                user = UserRepository.create_user(user_data)
            
            # Generate token
            token = cls.generate_jwt_token(user._id)
            
            return {
                'user': user.to_dict(),
                'token': token
            }
            
        except Exception as e:
            raise ValueError(f'Google login failed: {str(e)}')
    
    @classmethod
    def facebook_login(cls, facebook_token):
        """Login with Facebook OAuth"""
        try:
            # Verify Facebook token
            facebook_app_id = os.getenv('FACEBOOK_APP_ID')
            facebook_app_secret = os.getenv('FACEBOOK_APP_SECRET')
            
            if not facebook_app_id or not facebook_app_secret:
                raise ValueError('Facebook authentication not configured')
            
            # Verify token with Facebook
            verify_url = f'https://graph.facebook.com/debug_token?input_token={facebook_token}&access_token={facebook_app_id}|{facebook_app_secret}'
            response = requests.get(verify_url)
            
            if response.status_code != 200:
                raise ValueError('Invalid Facebook token')
            
            token_data = response.json()
            if not token_data.get('data', {}).get('is_valid'):
                raise ValueError('Invalid Facebook token')
            
            # Get user info from Facebook
            user_info_response = requests.get(
                f'https://graph.facebook.com/me?fields=id,name,email,picture&access_token={facebook_token}'
            )
            
            if user_info_response.status_code != 200:
                raise ValueError('Failed to get user info from Facebook')
            
            user_info = user_info_response.json()
            
            email = user_info.get('email', '').lower()
            name = user_info.get('name', '')
            facebook_id = user_info.get('id')
            profile_picture = user_info.get('picture', {}).get('data', {}).get('url')
            
            if not email or not facebook_id:
                raise ValueError('Invalid Facebook user data')
            
            # Check if user exists
            user = UserRepository.find_by_email(email)
            
            if user:
                # Update existing user
                update_data = {
                    'facebook_id': facebook_id,
                    'profile_picture': profile_picture,
                    'last_login': datetime.utcnow()
                }
                UserRepository.update_user(user._id, update_data)
                
                # Refresh user data
                user = UserRepository.find_by_id(user._id)
            else:
                # Create new user
                user_data = {
                    'email': email,
                    'name': name,
                    'facebook_id': facebook_id,
                    'provider': 'facebook',
                    'profile_picture': profile_picture,
                    'is_active': True,
                    'last_login': datetime.utcnow(),
                    'business_info': {
                        'business_name': name,
                        'business_type': 'Du lịch',
                        'industry': 'Tourism',
                        'phone': '',
                        'address': ''
                    }
                }
                
                user = UserRepository.create_user(user_data)
            
            # Generate token
            token = cls.generate_jwt_token(user._id)
            
            return {
                'user': user.to_dict(),
                'token': token
            }
            
        except Exception as e:
            raise ValueError(f'Facebook login failed: {str(e)}')
    
    @classmethod
    def get_user_by_token(cls, token):
        """Get user by JWT token"""
        user_id = cls.verify_jwt_token(token)
        user = UserRepository.find_by_id(user_id)
        
        if not user:
            raise ValueError('User not found')
        
        if not user.is_active:
            raise ValueError('Account is deactivated')
        
        return user
    
    @classmethod
    def update_user_profile(cls, user_id, profile_data):
        """Update user profile"""
        allowed_fields = ['name', 'profile_picture']
        update_data = {}
        
        for field in allowed_fields:
            if field in profile_data:
                if field == 'name' and profile_data[field]:
                    update_data[field] = profile_data[field].strip()
                else:
                    update_data[field] = profile_data[field]
        
        if not update_data:
            raise ValueError('No valid data to update')
        
        success = UserRepository.update_user(user_id, update_data)
        if not success:
            raise ValueError('Failed to update profile')
        
        # Return updated user
        return UserRepository.find_by_id(user_id)

File: be/src/auth/auth.service.ts
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import type { Prisma } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { toWireAdmin, toWireUser, type WireAdmin, type WireUser } from '../common/wire';
import type { RegisterDto, UpdateProfileDto } from './dto';

/**
 * Xac thuc nguoi dung va admin.
 *
 * MOT kieu hash duy nhat: bcrypt. Ban Python co ba kieu song song -
 * werkzeug (models/user.py), SHA-256 + salt (services/admin_auth_service.py)
 * va bcrypt (services/user_management_service.py). SHA-256 cho mat khau la
 * sai vi no qua nhanh nen brute-force re; bcrypt cham co chu dich.
 * Database dang trong nen thong nhat lai khong ton gi.
 */
@Injectable()
export class AuthService {
  // 12 vong: cham du de chong brute-force ma van duoi 100ms tren may thuong.
  private static readonly BCRYPT_ROUNDS = 12;

  constructor(
    private readonly db: PrismaService,
    private readonly jwt: JwtService,
  ) {}

  private sign(subject: string, type: 'user' | 'admin'): string {
    // Giu ten claim "user_id" giong ban Python de token cu con doc duoc
    // trong giai doan cat, va them "type" de phan biet user voi admin -
    // ban cu phai thu giai ma hai lan moi biet.
    return this.jwt.sign({ user_id: subject, type });
  }

  async register(dto: RegisterDto): Promise<{ user: WireUser; token: string }> {
    const email = dto.email.toLowerCase().trim();

    const existing = await this.db.user.findUnique({ where: { email } });
    if (existing) {
      // Cung thong diep voi ban cu de frontend hien nguyen van.
      throw new Error('User with this email already exists');
    }

    const user = await this.db.user.create({
      data: {
        email,
        name: dto.name.trim(),
        passwordHash: await bcrypt.hash(dto.password, AuthService.BCRYPT_ROUNDS),
        provider: 'email',
        // Ban cu tu dien businessInfo mac dinh khi khong duoc gui
        // (models/user.py duong 22). Giu nguyen hanh vi do.
        // Cast o bien gioi Prisma: Record<string, unknown> khong tu khop
        // InputJsonValue vi Prisma phan biet undefined voi JsonNull.
        businessInfo: (dto.businessInfo ?? {
          business_name: dto.name.trim(),
          business_type: 'Du lịch',
          industry: 'Tourism',
          phone: '',
          address: '',
        }) as Prisma.InputJsonValue,
      },
    });

    return { user: toWireUser(user), token: this.sign(user.id, 'user') };
  }

  async login(email: string, password: string): Promise<{
    user: WireUser | WireAdmin;
    token: string;
    isAdmin: boolean;
  }> {
    const key = email.toLowerCase().trim();

    // Thu admin truoc roi moi den user - giu dung thu tu cua ban Flask
    // (be/auth/auth.py duong 90), vi cung mot email co the ton tai o ca hai
    // bang va Dashboard dua vao viec admin duoc uu tien.
    const admin = await this.db.admin.findUnique({ where: { email: key } });
    if (admin?.isActive && (await bcrypt.compare(password, admin.passwordHash))) {
      // Dung KET QUA cua update, khong dung `admin` doc truoc do: neu tra ve
      // object cu thi response luon co last_login = null du DB da ghi dung.
      const fresh = await this.db.admin.update({
        where: { id: admin.id },
        data: { lastLoginAt: new Date() },
      });
      return { user: toWireAdmin(fresh), token: this.sign(fresh.id, 'admin'), isAdmin: true };
    }

    const user = await this.db.user.findUnique({ where: { email: key } });
    if (!user || !user.isActive || !user.passwordHash) {
      throw new UnauthorizedException('Invalid email or password');
    }
    if (!(await bcrypt.compare(password, user.passwordHash))) {
      throw new UnauthorizedException('Invalid email or password');
    }

    const fresh = await this.db.user.update({
      where: { id: user.id },
      data: { lastLoginAt: new Date() },
    });
    return { user: toWireUser(fresh), token: this.sign(fresh.id, 'user'), isAdmin: false };
  }

  /** Nguoi dung hoac admin ung voi token da giai ma. */
  async resolveSubject(
    id: string,
    type: string,
  ): Promise<{ user: WireUser | WireAdmin; userType: 'user' | 'admin' }> {
    if (type === 'admin') {
      const a = await this.db.admin.findUnique({ where: { id } });
      if (!a || !a.isActive) throw new UnauthorizedException('Admin not found or inactive');
      return { user: toWireAdmin(a), userType: 'admin' };
    }
    const u = await this.db.user.findUnique({ where: { id } });
    if (!u || !u.isActive) throw new UnauthorizedException('User not found or inactive');
    return { user: toWireUser(u), userType: 'user' };
  }

  async googleLogin(idToken: string): Promise<{ user: WireUser; token: string }> {
    const clientId = process.env.GOOGLE_CLIENT_ID;
    if (!clientId) throw new UnauthorizedException('Google authentication not configured');

    const res = await fetch(
      `https://oauth2.googleapis.com/tokeninfo?id_token=${encodeURIComponent(idToken)}`,
    );
    if (!res.ok) throw new UnauthorizedException('Invalid Google token');

    const info = (await res.json()) as Record<string, string>;
    // Kiem audience: khong kiem thi bat ky ID token Google nao cung dang
    // nhap duoc, ke ca token phat cho ung dung khac.
    if (info.aud !== clientId) throw new UnauthorizedException('Invalid token audience');

    const email = (info.email ?? '').toLowerCase();
    if (!email || !info.sub) throw new UnauthorizedException('Invalid Google user data');

    return this.upsertSocial({
      email,
      name: info.name ?? email,
      provider: 'google',
      googleId: info.sub,
      profilePicture: info.picture ?? null,
    });
  }

  async facebookLogin(accessToken: string): Promise<{ user: WireUser; token: string }> {
    const appId = process.env.FACEBOOK_APP_ID;
    const appSecret = process.env.FACEBOOK_APP_SECRET;
    if (!appId || !appSecret) {
      throw new UnauthorizedException('Facebook authentication not configured');
    }

    const debugRes = await fetch(
      `https://graph.facebook.com/debug_token?input_token=${encodeURIComponent(accessToken)}` +
        `&access_token=${encodeURIComponent(`${appId}|${appSecret}`)}`,
    );
    if (!debugRes.ok) throw new UnauthorizedException('Invalid Facebook token');
    const debug = (await debugRes.json()) as { data?: { is_valid?: boolean; app_id?: string } };
    if (!debug.data?.is_valid) throw new UnauthorizedException('Invalid Facebook token');
    // Ban Python khong kiem app_id tra ve. Kiem o day de token cua ung dung
    // Facebook khac khong dung duoc.
    if (debug.data.app_id && debug.data.app_id !== appId) {
      throw new UnauthorizedException('Token issued for another application');
    }

    const meRes = await fetch(
      `https://graph.facebook.com/me?fields=id,name,email,picture&access_token=${encodeURIComponent(accessToken)}`,
    );
    if (!meRes.ok) throw new UnauthorizedException('Failed to read Facebook profile');
    const me = (await meRes.json()) as {
      id?: string;
      name?: string;
      email?: string;
      picture?: { data?: { url?: string } };
    };
    const email = (me.email ?? '').toLowerCase();
    if (!email || !me.id) {
      // Facebook cho phep tai khoan khong co email; ban cu cung that bai o
      // day nhung khong noi ro ly do.
      throw new UnauthorizedException('Facebook account has no email address');
    }

    return this.upsertSocial({
      email,
      name: me.name ?? email,
      provider: 'facebook',
      facebookId: me.id,
      profilePicture: me.picture?.data?.url ?? null,
    });
  }

  /** Tao user moi hoac gan id social vao user da co cung email. */
  private async upsertSocial(p: {
    email: string;
    name: string;
    provider: 'google' | 'facebook';
    googleId?: string;
    facebookId?: string;
    profilePicture: string | null;
  }): Promise<{ user: WireUser; token: string }> {
    const link = p.provider === 'google' ? { googleId: p.googleId } : { facebookId: p.facebookId };

    const user = await this.db.user.upsert({
      where: { email: p.email },
      // Da co tai khoan cung email: gan id social vao thay vi tao ban trung.
      update: { ...link, lastLoginAt: new Date(), profilePicture: p.profilePicture },
      create: {
        email: p.email,
        name: p.name,
        provider: p.provider,
        profilePicture: p.profilePicture,
        lastLoginAt: new Date(),
        ...link,
      },
    });

    return { user: toWireUser(user), token: this.sign(user.id, 'user') };
  }

  async updateProfile(userId: string, dto: UpdateProfileDto): Promise<WireUser> {
    const user = await this.db.user.update({
      where: { id: userId },
      data: {
        ...(dto.name !== undefined ? { name: dto.name.trim() } : {}),
        ...(dto.profile_picture !== undefined ? { profilePicture: dto.profile_picture } : {}),
        ...(dto.businessInfo !== undefined
          ? { businessInfo: dto.businessInfo as Prisma.InputJsonValue }
          : {}),
      },
    });
    return toWireUser(user);
  }
}
